### PR TITLE
EDM-1271: refactor fleet selector task and remove overlapping selectors

### PR DESCRIPTION
**Pull request by avishayt**
**First commit:**
- Removed FleetOverlappingSelectors condition type and complex overlapping selector logic
- Simplified fleet selector task from 4 complex cases to 2 streamlined cases (device updates and fleet updates)
- Added new event types for device ownership changes
- Restructured task to handle device ownership transitions more efficiently
- Enhanced integration tests with complete event validation coverage
- Added ListDevicesByServiceCondition method for querying devices by service conditions
- Improved error handling and logging throughout the fleet selector pipeline

This refactoring removes the complexity of overlapping selector detection and replaces it with a simpler, more efficient approach that directly manages device ownership and emits appropriate events for all state transitions.

**Second commit:**
Standardize event messages for consistency and clarity
- All messages now start with uppercase letters
- Complete sentences with proper punctuation (end with periods)
- Consistent past tense for actions (e.g., 'was created', 'was updated')
- Include relevant context (resource name/type with quotes)

**Third commit:**
Create gin index for device service conditions
- Also separate each index to its own method to reduce cyclomatic complexity.
    
**WARNING: This commit changes the service_conditions column to be jsonb which is a schema change.**

## Summary by CodeRabbit

* **New Features**
  * Introduced new event types and detailed event data for device ownership changes, detection and resolution of multiple owners, internal task failures, and fleet selector processing completion.
  * Added API and CLI support to list devices by specific service conditions, such as devices with multiple owners.
  * Enhanced event emission and observability for fleet selector and device ownership operations.

* **Improvements**
  * Refined device ownership conflict handling: conflicts are now indicated on devices (as "MultipleOwners") instead of fleets.
  * Streamlined and unified fleet selector matching logic for improved reliability, error handling, and observability.
  * Updated documentation to reflect the new conflict condition handling and usage.
  * Improved status message text consistency with updated punctuation.
  * Simplified fleet selector task submission logic by unifying operation types.
  * Enhanced event message formatting and standardized lifecycle event messages.

* **Bug Fixes**
  * Enhanced error handling and reporting during fleet and device updates.

* **Tests**
  * Expanded integration and end-to-end tests to validate new event emissions, ownership logic, pagination, context cancellation, and edge cases.
  * Updated tests to remove deprecated overlapping selector condition checks and validate new event-driven ownership logic.

### CODE REVIEW

File: internal/store/device.go [hunk 1]
```diff
@@ -92,20 +95,53 @@ func (s *DeviceStore) InitialMigration(ctx context.Context) error {
...
+	return nil
+}
+
+func (s *DeviceStore) createDeviceNameIndex(db *gorm.DB) error {
 	if !db.Migrator().HasIndex(&model.Device{}, "idx_device_primary_key_name") {
 		if db.Dialector.Name() == "postgres" {
-			if err := db.Exec("CREATE INDEX idx_device_primary_key_name ON devices USING BTREE (name)").Error; err != nil {
-				return err
-			}
+			return db.Exec("CREATE INDEX idx_device_primary_key_name ON devices USING BTREE (name)").Error
```
Comment (gshilin-sdb) [inline internal/store/device.go:136]: shouldn't all of "CREATE INDEX" to be "IF NOT EXISTS "?

File: internal/store/device.go [hunk 2]
```diff
@@ -683,3 +722,71 @@ func (s *DeviceStore) GetRepositoryRefs(ctx context.Context, orgId uuid.UUID, na
...
+func (s *DeviceStore) ListDevicesByServiceCondition(ctx context.Context, orgId uuid.UUID, conditionType string, conditionStatus string, listParams ListParams) (*api.DeviceList, error) {
+	// Use raw SQL to efficiently query JSONB service_conditions field
+	// The service_conditions field is stored as JSONB for optimal performance
+	var devices []model.Device
+	var nextContinue *string
+	var numRemaining *int64
+
+	// Build the raw SQL query with proper pagination support for JSONB
+	baseSQL := `
+		SELECT * FROM devices
```
Comment (gshilin-sdb) [inline internal/store/device.go:735]: should we ignore decommissioned devices?

File: internal/tasks/fleet_selector.go [hunk 3]
```diff
@@ -319,191 +355,175 @@ func (f FleetSelectorMatchingLogic) removeOwnerFromMatchingDevices(ctx context.C
...
-	condErrors := 0
-
-	// overlappingFleets acts as a set of strings
-	overlappingFleets := map[string]struct{}{}
+		for _, device := range devices.Items {
+			// Check for context cancellation in long-running loops
+			if ctx.Err() != nil {
+				f.log.Warnf("Context cancelled during device processing, stopping early. Processed %d devices so far", devicesProcessed)
+				errors++
+				return devicesProcessed, errors
```
Comment (gshilin-sdb) [inline internal/tasks/fleet_selector.go:398]: maybe `return devicesProcessed, errors + 1` without `errors++`?

File: internal/tasks/fleet_selector.go [hunk 4]
```diff
@@ -660,3 +671,214 @@ func findMatchingFleets(labels map[string]string, fleets []api.Fleet) []string {
...
+	for k, v := range labelsMap {
+		keys = append(keys, k)
+		values = append(values, v)
+	}
+
+	// Construct selector: devices that don't match AND are owned by this fleet
+	ls := fmt.Sprintf("(%s) != (%s)", strings.Join(keys, ","), strings.Join(values, ","))
+	listParams := api.ListDevicesParams{
+		Limit:         lo.ToPtr(f.itemsPerPage),
+		LabelSelector: lo.ToPtr(ls),
```
Comment (gshilin-sdb) [inline internal/tasks/fleet_selector.go]: why not use fmt.Sprintf directly without special variable, like it's done on the next line?

File: internal/tasks/fleet_selector.go [hunk 5]
```diff
@@ -660,3 +671,214 @@ func findMatchingFleets(labels map[string]string, fleets []api.Fleet) []string {
...
+	devicesProcessed, errors := 0, 0
+
+	// Use store-level pagination - start with empty continue
+	listParams := store.ListParams{Limit: int(f.itemsPerPage)}
+
+	for {
+		// Check for context cancellation in long-running loops
+		if ctx.Err() != nil {
+			f.log.Warnf("Context cancelled during multiple owners condition processing, stopping early. Processed %d devices so far", devicesProcessed)
+			return devicesProcessed, errors
```
Comment (gshilin-sdb) [inline internal/tasks/fleet_selector.go:796]: Did we count this error already? I mean shouldn't it be `errors + 1`?

File: internal/tasks/fleet_selector.go [hunk 6]
```diff
@@ -660,3 +671,214 @@ func findMatchingFleets(labels map[string]string, fleets []api.Fleet) []string {
...
+
+		if len(devices.Items) == 0 {
+			break // No more devices to process
+		}
+
+		for _, device := range devices.Items {
+			// Check for context cancellation in long-running loops
+			if ctx.Err() != nil {
+				f.log.Warnf("Context cancelled during device processing, stopping early. Processed %d devices so far", devicesProcessed)
+				return devicesProcessed, errors
```
Comment (gshilin-sdb) [inline internal/tasks/fleet_selector.go:814]: Shouldn't it be `errors + 1` too?

File: internal/tasks/fleet_selector.go [hunk 7]
```diff
@@ -660,3 +671,214 @@ func findMatchingFleets(labels map[string]string, fleets []api.Fleet) []string {
...
+			}
+
+			devicesProcessed++
+			// Recompute matching fleets for this orphaned device
+			allFleets, err := fetcher()
```
Comment (gshilin-sdb) [inline internal/tasks/fleet_selector.go]: maybe to rename `fetcher()` fo `fleetFetcher()` everywhere? Or even `allFleetsFetcher`?

File: internal/service/event.go [hunk 8]
```diff
@@ -246,3 +306,152 @@ func GetResourceEventFromUpdateDetails(ctx context.Context, resourceKind api.Res
...
+	message := formatDeviceOwnershipMessage(previousOwner, newOwner)
+
+	details := api.EventDetails{}
+	detailsStruct := api.DeviceOwnershipChangedDetails{
+		PreviousOwner: previousOwner,
+		NewOwner:      newOwner,
+	}
+	if err := details.FromDeviceOwnershipChangedDetails(detailsStruct); err != nil {
+		log.WithError(err).Error("Failed to serialize device ownership changed event details")
+		return nil
```
Comment (gshilin-sdb) [inline internal/service/event.go:320]: In line 210 we preferred to emit event in any case: `// Ignore the error and create an event, even without details`
To be consistent we should do the same here and in the following functions.

File: internal/service/event.go [hunk 9]
```diff
@@ -246,3 +306,152 @@ func GetResourceEventFromUpdateDetails(ctx context.Context, resourceKind api.Res
...
+	}
+
+	var status api.Status
+	var reason api.EventReason
+	var outcomeFunc func() string
+
+	if devicesWithErrors > 0 {
+		status = api.StatusInternalServerError("Fleet selector processing completed with errors")
+		reason = api.EventReasonFleetSelectorProcessingCompleted
```
Comment (gshilin-sdb) [inline internal/service/event.go]: do we need these two variables?

File: internal/store/device.go [hunk 10]
```diff
@@ -683,3 +722,71 @@ func (s *DeviceStore) GetRepositoryRefs(ctx context.Context, orgId uuid.UUID, na
...
+	if err := s.getDB(ctx).Raw(baseSQL, args...).Scan(&devices).Error; err != nil {
+		return nil, ErrorFromGormError(err)
+	}
+
+	// Calculate pagination metadata
+	if len(devices) > 0 && len(devices) == listParams.Limit {
+		// Check if there are more results
+		var count int64
+		countSQL := `
+			SELECT COUNT(*) FROM devices
```
Comment (gshilin-sdb) [inline internal/store/device.go:770]: `SELECT COUNT(*)` is very slow in PostgreSQL. Especially here with a complex WHERE.
Probably that's too early to optimize, but...

File: internal/store/device.go [hunk 11]
```diff
@@ -117,70 +153,75 @@ func (s *DeviceStore) InitialMigration(ctx context.Context) error {
...
+			return db.Exec("CREATE INDEX IF NOT EXISTS idx_devices_service_conditions ON devices USING GIN ((service_conditions->'conditions')) WHERE service_conditions IS NOT NULL").Error
+		} else {
+			return db.Migrator().CreateIndex(&model.Device{}, "ServiceConditions")
+		}
+	}
+	return nil
+}
+
+func (s *DeviceStore) createDeviceLabelsTrigger(db *gorm.DB) error {
 	if db.Dialector.Name() == "postgres" {
```
Comment (gshilin-sdb) [inline internal/store/device.go:225]: do we really care about non-posgres DBs?

File: internal/service/event.go (hunk 8 above)
Comment (avishayt) [reply internal/service/event.go:320]: Fixed

File: internal/service/event.go (hunk 9 above)
Comment (avishayt) [reply internal/service/event.go]: Fixed

File: internal/store/device.go (hunk 1 above)
Comment (avishayt) [reply internal/store/device.go:136]: We already check using "if !db.Migrator().HasIndex()" above

File: internal/store/device.go (hunk 11 above)
Comment (avishayt) [reply internal/store/device.go:225]: We're not optimizing for other DBs, but it should work reasonably

File: internal/store/device.go (hunk 2 above)
Comment (avishayt) [reply internal/store/device.go:735]: This is a lower-level method that doesn't need to care.  Ignoring decommissioned devices is done by the task.

File: internal/store/device.go (hunk 10 above)
Comment (avishayt) [reply internal/store/device.go:770]: It's used in at least 2 other places.  We can optimize at another time.

File: internal/tasks/fleet_selector.go (hunk 3 above)
Comment (avishayt) [reply internal/tasks/fleet_selector.go:398]: I had this "return devicesProcessed, errors + 1" in another place, but I actually prefer this.  It makes it easier to understand IMO, separating the error count and the return.  I changed the other line.

Comment (avishayt) [reply internal/tasks/fleet_selector.go:398]: Actually CodeRabbit made that suggestion:
https://github.com/flightctl/flightctl/pull/1325/files#r2188485980

File: internal/tasks/fleet_selector.go (hunk 4 above)
Comment (avishayt) [reply internal/tasks/fleet_selector.go]: Done

File: internal/tasks/fleet_selector.go (hunk 7 above)
Comment (avishayt) [reply internal/tasks/fleet_selector.go]: Yes, agree

File: internal/tasks/fleet_selector.go (hunk 5 above)
Comment (avishayt) [reply internal/tasks/fleet_selector.go:796]: No, context cancellation is not an error

File: internal/tasks/fleet_selector.go (hunk 6 above)
Comment (avishayt) [reply internal/tasks/fleet_selector.go:814]: No, context cancellation is not an error

File: internal/store/device.go (hunk 11 above)
Comment (gshilin-sdb) [reply internal/store/device.go:225]: Can we use postgres-only stuff, like CTE?

Comment (avishayt) [reply internal/store/device.go:225]: Sure, but we should try to have a fallback for non-postgres

**Review by gshilin-sdb — APPROVED** [review]
